service/qos/qos_table.cc

```diff
@@ -119,4 +120,39 @@ future<service_levels_info> get_service_level(cql3::query_processor& qp, const s
     });
 }
 
+static data_value timeout_to_data_value(const qos::service_level_options::timeout_type& tv) {
+    return std::visit(overloaded_functor {
+        [&] (const qos::service_level_options::unset_marker&) {
+            return data_value::make_null(duration_type);
+        },
+        [&] (const qos::service_level_options::delete_marker&) {
+            return data_value::make_null(duration_type);
+        },
+        [&] (const lowres_clock::duration& d) {
+            return data_value(cql_duration(months_counter{0},
+                    days_counter{0},
+                    nanoseconds_counter{std::chrono::duration_cast<std::chrono::nanoseconds>(d).count()}));
+        },
+    }, tv);
+}
+
+future<std::vector<mutation>> get_set_mutations(cql3::query_processor& qp, api::timestamp_type timestamp, sstring service_level_name, qos::service_level_options slo) {
+    static sstring insert_query = format("INSERT INTO {}.{} (service_level) VALUES (?);", db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2);
+    static sstring update_query = format("UPDATE {}.{} SET timeout = ?, workload_type = ? WHERE service_level = ?;", db::system_keyspace::NAME, db::system_keyspace::SERVICE_LEVELS_V2);
+    data_value workload = slo.workload == qos::service_level_options::workload_type::unspecified
+            ? data_value::make_null(utf8_type)
+            : data_value(qos::service_level_options::to_string(slo.workload));
+    
+    auto insert_mutations = co_await qp.get_mutations_internal(insert_query, internal_query_state(), timestamp, {service_level_name});
+    auto update_mutations = co_await qp.get_mutations_internal(update_query, internal_query_state(), timestamp, {timeout_to_data_value(slo.timeout), workload, service_level_name});
+    insert_mutations.insert(insert_mutations.end(), update_mutations.begin(), update_mutations.end());
+
```

Comment:
Why not use insert unconditionally?

Reply:
You use insert_query for part of the update and update_query for the other. Can't it be one INSERT?